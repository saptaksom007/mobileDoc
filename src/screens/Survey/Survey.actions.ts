export const ActionTypes = {
  SELECT_SURVEY_REQUESTED: 'SELECT_SURVEY_REQUESTED',
  SELECT_SURVEY_SUCCEEDED: 'SELECT_SURVEY_SUCCEEDED',
  SELECT_SURVEY_FAILED: 'SELECT_SURVEY_FAILED'
}

export default class Actions {
  static selectSurveyRequested(url: string, patientId: string) {
    return {
      type: ActionTypes.SELECT_SURVEY_REQUESTED,
      payload: {
        url,
        patientId
      }
    }
  }
  static selectSurveySucceeded() {
    return { type: ActionTypes.SELECT_SURVEY_SUCCEEDED }
  }
  static selectSurveyFailed(error: Error) {
    return {
      type: ActionTypes.SELECT_SURVEY_FAILED,
      message: error.message,
      error
    }
  }
}
