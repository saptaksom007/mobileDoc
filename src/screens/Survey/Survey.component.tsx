import React, { Component } from 'react'

import { ScrollView, InteractionManager, View } from 'react-native'
import { ListItem, List } from 'react-native-elements'
import { connect } from 'react-redux'
import isString from 'lodash/isString'
import i18n from 'ex-react-native-i18n'
import { Paragraph } from 'components/CustomText/CustomText.component'
import { surveyActions } from 'common-docdok/lib/domain/survey/actions'
import { findPatientSelf } from 'common-docdok/lib/domain/healthrelation/selectors/findPatient'

import { compose } from 'recompose'
import { propOr, pipe, find } from 'ramda'
import { NavigationScreenProp } from 'react-navigation'

import { Actions as NavAction } from 'navigation/SagaNavigation'
import NotificationsActions from 'notifications/actions'
import EmptyList from 'components/EmptyList/EmptyList.component'
import withSubtitle from 'screens/enhancers/withSubtitle'
import GraphicActions from 'screens/Graphic/Graphic.actions'

import { momentWithLocales } from 'utilities/date'
import toString from 'common-docdok/lib/utils/toString'

import { Color } from 'constants/Color'
import { Env } from 'env'

import Actions from './Survey.actions'
import styles from './Survey.styles'

function getSubtitle(survey: any) {
  const datePattern = 'DD.MM.YY, hh:mm A'

  const creationDate = momentWithLocales(survey.creationDate).format(
    datePattern,
  )
  return (
    <View style={{ marginLeft: 10, marginTop: 5 }}>
      <Paragraph style={{ color: '#999' }}>{`${i18n.t(
        'surveys.sent',
      )} ${creationDate}`}</Paragraph>
      <Paragraph style={{ color: '#999' }}>{`${
        !survey.completionDate
          ? i18n.t('surveys.notComplete')
          : i18n.t('surveys.complete') +
            ' - ' +
            momentWithLocales(survey.completionDate).format(datePattern)
      }`}</Paragraph>
    </View>
  )
}

function getTitle(survey: any, doctor: any) {
  const surveyName = propOr<string, any, string>(
    i18n.t('surveys.default.surveyName'),
    'surveyName',
    survey,
  )
  return `${surveyName}${surveyName.length > 50 ? '\n' : ' '}${i18n.t(
    'surveys.from',
  )} ${toString.person(doctor)}`
}

interface Props {
  dispatch(action: any): void
  mySurveys: any[]
  users: { [key: string]: any }
  navigation: NavigationScreenProp<any, any>
  patientSelfId?: string
}
const filterPatient = (patientId: string) => (survey: any) =>
  survey.patientUuid === patientId

export class Survey extends Component<Props> {
  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      const { dispatch } = this.props
      dispatch(surveyActions.getMySurveysRequested())
      dispatch(NotificationsActions.removeNotificationRequested('SURVEY'))
    })
  }

  selectSurvey(survey: any) {
    const {
      dispatch,
      navigation: {
        state: {
          params: { patientId },
        },
      },
    } = this.props
    if (survey.completionDate) {
      if (__DEV__) {
        this.props.dispatch(
          GraphicActions.graphicRequested(Number(survey.surveyId), patientId),
        )
      }
      dispatch(
        NavAction.showLocalWarning(i18n.t('surveys.alert.completionDate')),
      )
    } else {
      const url = `${Env.api.base}/private/surveyparticipation/${survey.id}`
      dispatch(Actions.selectSurveyRequested(url, patientId))
    }
  }

  render() {
    const {
      mySurveys,
      users,
      dispatch,
      navigation: {
        state: { params },
      },
      patientSelfId,
    } = this.props

    const surveys =
      (
        mySurveys &&
        mySurveys.filter(filterPatient(params.patientId || patientSelfId))
      ).sort(
        (s1, s2) =>
          new Date(s2.creationDate || 0).getTime() -
          new Date(s1.creationDate || 0).getTime(),
      ) || []

    if (!surveys || surveys.length === 0) {
      return (
        <EmptyList
          text={i18n.t('surveys.empty')}
          show
          onPress={() => dispatch(surveyActions.getMySurveysRequested())}
        />
      )
    }

    return (
      <ScrollView style={styles.container}>
        <List containerStyle={styles.listContainer}>
          {surveys.map((survey: any) => (
            <ListItem
              key={survey.id}
              title={getTitle(survey, users[survey.senderRef])}
              titleNumberOfLines={4}
              containerStyle={styles.itemContainerStyle}
              titleStyle={styles.itemTitleStyle}
              subtitle={getSubtitle(survey)}
              subtitleStyle={styles.itemSubtitleStyle}
              onPress={() => this.selectSurvey(survey)}
              hideChevron={isString(survey.completionDate)}
              chevronColor={Color.tintColor}
            />
          ))}
        </List>
      </ScrollView>
    )
  }
}

function select({
  survey: { mySurveys },
  healthrelation: { patients },
  userCache: { users },
}: any) {
  const patientSelfId = pipe(
    find(findPatientSelf),
    propOr('', 'uuid'),
  )(patients)
  return { mySurveys, users, patientSelfId }
}

export default compose<Props, any>(
  connect(select),
  withSubtitle('surveys.subtitle'),
)(Survey)
