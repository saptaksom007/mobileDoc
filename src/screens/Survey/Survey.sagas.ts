import { put, takeEvery } from 'redux-saga/effects'

import { Actions as NavAction } from 'navigation/SagaNavigation'
import i18n from 'ex-react-native-i18n'

import Actions, { ActionTypes } from './Survey.actions'

function* gotoSurveyWebView({ payload: { url, patientId } }: any) {
  try {
    yield put(NavAction.push('surveywebview', { url, patientId }))
    yield put(Actions.selectSurveySucceeded())
  } catch (err) {
    yield put(Actions.selectSurveyFailed(err))
    yield put(
      NavAction.showLocalError(i18n.t('surveys.errors.gotoSurveyWebView')),
    )
  }
}

export default [
  takeEvery(ActionTypes.SELECT_SURVEY_REQUESTED, gotoSurveyWebView),
]
