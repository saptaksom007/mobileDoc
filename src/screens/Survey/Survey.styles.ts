import { StyleSheet } from 'react-native'
import { Color } from 'constants/Color'
import {
  fontFamilyBold,
  fontFamily
} from 'components/CustomText/CustomText.styles'

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff'
  },
  listContainer: {
    backgroundColor: '#fff',
    marginTop: 0,
    borderTopWidth: 0,
    borderBottomWidth: 0
  },
  itemContainerStyle: { borderBottomColor: Color.borderColor },
  itemTitleStyle: { color: Color.darkText, fontFamily: fontFamilyBold },
  itemSubtitleStyle: { color: Color.lightText, fontFamily }
})
