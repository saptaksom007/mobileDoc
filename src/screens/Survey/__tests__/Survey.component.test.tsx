import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { Survey } from '../Survey.component'
import { NavigationScreenProp } from 'react-navigation'

it('<Survey/> render correctly', () => {
  const tree = renderer
    .create(
      <Survey
        users={{
          123: {}
        }}
        dispatch={() => {}}
        mySurveys={[]}
        navigation={{ state: { params: {} } } as NavigationScreenProp<any, any>}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Survey/> render correctly', () => {
  const tree = renderer
    .create(
      <Survey
        users={{
          123: {}
        }}
        dispatch={() => {}}
        mySurveys={[
          {
            senderRef: '123',
            completionDate: '',
            surveyName: 'Test survey',
            id: 1
          }
        ]}
        navigation={{ state: { params: {} } } as NavigationScreenProp<any, any>}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
