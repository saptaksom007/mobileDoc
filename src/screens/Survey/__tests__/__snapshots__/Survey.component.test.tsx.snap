// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Survey/> render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 1294,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Array [
        Object {
          "fontFamily": "TitilliumWeb-Regular",
          "fontSize": 17.5,
          "lineHeight": 22,
          "overflow": "visible",
        },
        Object {
          "color": "rgba(27, 106, 158, 0.85)",
          "fontSize": 17.5,
          "lineHeight": 21,
        },
        Object {
          "color": "#999",
          "marginTop": -40,
        },
      ]
    }
  >
    surveys.empty
  </Text>
</View>
`;

exports[`<Survey/> render correctly 2`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View />
</RCTScrollView>
`;
